### PR TITLE
CSI: Ability to disable snapshotter from ceph-csi rbd

### LINKED ISSUE #4401: Ability to disable snapshotter sidecar from the csi driver pod spec
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The CSI snapshotter feature is still in alpha and is therefore not available unless a quality gate is enabled. Still, the CSI driver always enables the snapshotter sidecar.

**Expected behavior:**
The operator should expose an option to disable the snapshotter sidecar at least until the snapshotter CSI spec is stable.

**How to reproduce it (minimal and precise):**

- Install the ceph operator
- Create a cluster
- See the pod spec for the csi driver always includes the snapshotter

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**

with this change, the snapshotter sidecar container won't be deployed by default with the rbd provisioner. we need to set a ENV `CSI_RBD_ENABLE_SNAPSHOTTER` variable in the operator template to enable deployment of snapshot sidecar in ceph-csi rbd provisioner.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #4401

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.
[test ceph]

### CODE REVIEW

File: pkg/operator/ceph/csi/spec.go [hunk 1]
```diff
@@ -199,6 +202,11 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 	tp.RBDGRPCMetricsPort = getPortFromENV("CSI_RBD_GRPC_METRICS_PORT", DefaultRBDGRPCMerticsPort)
 	tp.RBDLivenessMetricsPort = getPortFromENV("CSI_RBD_LIVENESS_METRICS_PORT", DefaultRBDLivenessMerticsPort)
 
+	enableSnap := os.Getenv("CSI_RBD_ENABLE_SNAPSHOTTER")
+
+	if strings.ToLower(enableSnap) == "true" {
```
Comment (travisn) [inline pkg/operator/ceph/csi/spec.go]: For case-insensitive compare in golang, you can use `strings.EqualFold()`.

**Comment by Madhu-1** [comment]
@travisn updated PR  PTAL

File: pkg/operator/ceph/csi/spec.go [hunk 2]
```diff
@@ -199,6 +202,11 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 	tp.RBDGRPCMetricsPort = getPortFromENV("CSI_RBD_GRPC_METRICS_PORT", DefaultRBDGRPCMerticsPort)
 	tp.RBDLivenessMetricsPort = getPortFromENV("CSI_RBD_LIVENESS_METRICS_PORT", DefaultRBDLivenessMerticsPort)
 
+	enableSnap := os.Getenv("CSI_RBD_ENABLE_SNAPSHOTTER")
+
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go]: You don't need this empty line.

File: pkg/operator/ceph/csi/spec.go [hunk 3]
```diff
@@ -349,7 +357,7 @@ func createCSIDriverInfo(clientset kubernetes.Interface, name string) error {
 		return nil
 	}
 	if apierrors.IsAlreadyExists(err) {
-		logger.Info("CSIDriver CRD already had been registered for %q", name)
+		logger.Infof("CSIDriver CRD already had been registered for %q", name)
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go]: This should be a separate commit since this has nothing to do with the original goal of the PR.

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -199,6 +202,11 @@ func StartCSIDrivers(namespace string, clientset kubernetes.Interface, ver *vers
 	tp.RBDGRPCMetricsPort = getPortFromENV("CSI_RBD_GRPC_METRICS_PORT", DefaultRBDGRPCMerticsPort)
 	tp.RBDLivenessMetricsPort = getPortFromENV("CSI_RBD_LIVENESS_METRICS_PORT", DefaultRBDLivenessMerticsPort)
 
+	enableSnap := os.Getenv("CSI_RBD_ENABLE_SNAPSHOTTER")
+
+	if !strings.EqualFold(enableSnap, "false") {
+		tp.EnableSnapshotter = "true"
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go]: Shouldn't this be decoupled into both rbd and cephfs?

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: the main reason to disable the snapshot is due to the alpha feature, this holds good for both rbd and cephfs 
updated the env variable name to support both cephfs and rbd

File: pkg/operator/ceph/csi/spec.go (hunk 3 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: done

**Comment by Madhu-1** [comment]
@leseb updated PTAL

File: pkg/operator/ceph/csi/spec.go (hunk 2 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go]: done

File: pkg/operator/ceph/csi/spec.go (hunk 4 above)
Comment (leseb) [reply pkg/operator/ceph/csi/spec.go]: > the main reason to disable the snapshot is due to the alpha feature, this holds good for both rbd and cephfs


> updated the env variable name to support both cephfs and rbd

Where is this?

Comment (leseb) [reply pkg/operator/ceph/csi/spec.go]: Oh ok, now we have a more general variable. Sounds good!

**Comment by Madhu-1** [comment]
@leseb @travisn CI passed can we merge this one?
